editor/src/components/canvas/canvas-strategies/resize-absolute-strategy.ts

```diff
@@ -0,0 +1,473 @@
+    const drag = sessionProps.drag
+    if (drag != null) {
+      delta = Utils.scaleVector(
+        Utils.scalePoint(drag, resizeHandle.enabledDirection as CanvasVector),
+        deltaScale,
+      )
+    }
+    const startingCorner: EdgePosition = {
+      x: 1 - resizeHandle.edgePosition.x,
+      y: 1 - resizeHandle.edgePosition.y,
+    } as EdgePosition
+    const startingPoint = pickPointOnRect(originalSize, startingCorner)
+    const originalCenter = Utils.getRectCenter(originalSize)
+    const draggedCorner = pickPointOnRect(originalSize, resizeHandle.edgePosition)
+
+    const newCorner = Utils.offsetPoint(draggedCorner, delta)
+    const snappedNewCorner = Utils.roundPointTo(
+      snapPoint(
+        editor,
+        newCorner,
+        propertyChange.enableSnapping,
+        propertyChange.keepAspectRatio,
+        startingPoint,
+        draggedCorner,
+        startingCorner,
+      ),
+      0,
+    )
+    const newSizeVector = Utils.pointDifference(startingPoint, snappedNewCorner)
+    const newRectangle = propertyChange.centerBasedResize
+      ? Utils.rectFromPointVector(originalCenter, Utils.scaleVector(newSizeVector, 0.5), true)
+      : Utils.rectFromPointVector(startingPoint, newSizeVector, false)
+    return newRectangle
+  }
+}
+
+export function calculateNewBounds(
+  editor: EditorState,
+  resizeHandle: ResizeHandle,
+  sessionProps: SelectModeCanvasSessionProps,
+  sessionState: SelectModeCanvasSessionState,
+): CanvasRectangle {
+  const originalSize = resizeHandle.originalSize
+  const aspectRatio = originalSize.width / originalSize.height
+  const newRectangle = calculateDraggedRectangle(editor, resizeHandle, sessionProps, sessionState)
+  const resizeOptions = editor.canvas.resizeOptions
+
+  const propertyChange = findResizePropertyChange(
+    sessionState.absoluteResizeStrategy?.properties ?? [],
+    resizeOptions,
+  )
+  if (propertyChange == null) {
+    return originalSize
+  } else {
+    // At this point I do ugly things to keep side drags in line
+    if (resizeHandle.edgePosition.x === 0.5) {
+      const newWidth = propertyChange.keepAspectRatio
+        ? Utils.roundTo(newRectangle.height * aspectRatio)
+        : originalSize.width
+      newRectangle.x -= newWidth / 2
+      newRectangle.width = newWidth
+    }
+    if (resizeHandle.edgePosition.y === 0.5) {
+      const newHeight = propertyChange.keepAspectRatio
+        ? Utils.roundTo(newRectangle.width / aspectRatio)
+        : originalSize.height
+      newRectangle.y -= newHeight / 2
+      newRectangle.height = newHeight
+    }
+
+    return newRectangle
+  }
+}
+
+function applyResize(
+  jsxMetadata: ElementInstanceMetadataMap,
+  frameAndTarget: PinFrameChange,
+  element: JSXElement,
+): Array<ValueAtPath> {
+  let propsToSet: Array<ValueAtPath> = []
+  let propsToSkip: Array<PropertyPath> = []
+  const nonGroupParent = MetadataUtils.findParent(jsxMetadata, frameAndTarget.target)
+  const parentFrame =
```

Comment:
TODO: we have `specialSizeMeasurements.immediateParentBounds` that is the "bounds to use" as determined by the dom-walker, and handles some edge cases this code doesn't. (like if the parent is a component with hierarchy itself)